Source/WebCore/platform/graphics/cg/CGUtilities.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "FloatRect.h"
+#include "GraphicsTypes.h"
+#include "IntRect.h"
+#include <CoreGraphics/CoreGraphics.h>
+#include <math.h>
+
+namespace WebCore {
+
+inline CGInterpolationQuality cgInterpolationQuality(InterpolationQuality quality)
+{
+    switch (quality) {
+    case InterpolationQuality::Default:
+        return kCGInterpolationDefault;
+    case InterpolationQuality::DoNotInterpolate:
+        return kCGInterpolationNone;
+    case InterpolationQuality::Low:
+        return kCGInterpolationLow;
+    case InterpolationQuality::Medium:
+        return kCGInterpolationMedium;
+    case InterpolationQuality::High:
+        return kCGInterpolationHigh;
+    }
+    ASSERT_NOT_REACHED();
+    return kCGInterpolationDefault;
+}
+
+inline FloatRect cgRoundToDevicePixelsNonIdentity(CGAffineTransform deviceMatrix, FloatRect rect)
```

Comment:
Does this function have to be CG only? Can't we pass AffineTransform instead and move non CG specific file?

Reply:
Yes, but should we do it in this patch or in some other patch?
(I can also rewrite the algorithm to read simpler, but should that be done in this patch or other patch?)